libcontainer/container_linux.go

```diff
@@ -523,10 +544,39 @@ func (c *linuxContainer) newInitProcess(p *Process, cmd *exec.Cmd, messageSockPa
 		}
 	}
 	_, sharePidns := nsMaps[configs.NEWPID]
-	data, err := c.bootstrapData(c.config.Namespaces.CloneFlags(), nsMaps)
+	data, err := c.bootstrapData(c.config.Namespaces.CloneFlags(), nsMaps, initStandard)
 	if err != nil {
 		return nil, err
 	}
+
+	if c.shouldSendMountSources() {
+		var mountFileFdsList strings.Builder
+
+		// We know the size of the string can easily len(c.config.Mounts). We multiply that by two, assuming
+		// the number we append has 2 digits top, to reduce allocations in most cases.
+		mountFileFdsList.Grow(len(c.config.Mounts) * 2)
+
+		for _, m := range c.config.Mounts {
+			if m.Device != "bind" {
+				// StartInitialization() finds out the Mounts indices by counting ";".
+				// We take care of adding the right number of ";"
+				mountFileFdsList.WriteString(";")
+				continue
+			}
+
+			// The fd passed here will not be used: nsexec.c will overwrite it with dup3(). We just need
+			// to allocate a fd so that we know the number to pass in the environment variable. The fd
+			// must not be closed before cmd.Start(), so we reuse messageSockPair.child because the
+			// lifecycle of that fd is already taken care of.
+			cmd.ExtraFiles = append(cmd.ExtraFiles, messageSockPair.child)
+			mountFileFdsList.WriteString(strconv.Itoa(stdioFdCount + len(cmd.ExtraFiles) - 1))
```

Comment:
Hmm, it seems that you forget to add `;` after the number here (or I'm grossly misunderstanding something).

Reply:
And this bug is not caught by the tests because in the tests you only have a single mount. Please amend the test case, see that it catches the bug, then fix this :)

Reply:
Yes, you are right on. I introduced this regression when switching to strings.Builder. Added a new test that fails with this code, and adding the ";" fixes the tests. Thanks for catching this!